src/core/coap/coap_client.cpp

```diff
@@ -435,15 +410,16 @@ void Client::ProcessReceivedMessage(Message &aMessage, const Ip6::MessageInfo &a
         break;
 
     case kCoapTypeConfirmable:
+        // Send empty ACK if it is a CON message.
+        SendAck(responseHeader, aMessageInfo);
+
     case kCoapTypeNonConfirmable:
-        if (responseHeader.IsConfirmable())
+        if (responseHeader.IsResponse() && responseHeader.IsTokenEqual(requestHeader))
```

Comment:
Actually the token was already checked in FindRelatedRequest request function. If it did not match for NON/CON messages, it would not find the message.